### ISSUE TITLE
Debugpy stops working after running `nc` to detect when its port is up

### ISSUE THREAD

**Issue by felipecrs** [microsoft/debugpy#974]
I faced the same issue today. And I realized that the problem was caused because I was performing some `nc -z <host> 5678` as a `preDebugTask` to ensure VS Code would only try to attach after the debugpy is up.

Immediately after stopping doing these `nc` calls and only placing `sleep <some seconds>`, it fixed the problem for me. Which is unfortunate, and I wonder if this could be a bug in Debugpy.

- Refs https://github.com/blakeblackshear/frigate-hass-integration/pull/302

_Originally posted by @felipecrs in https://github.com/microsoft/debugpy/issues/423#issuecomment-1177918816_

**Comment by fabioz** [comment]
Humm, I'm not sure about this... the backend will only accept a single connection, so, if you use `nc` to create such a connection, it's expected that a 2nd one won't be able to connect (so, this is working as intended right now).

Security-wise we don't want the debugger listening after the initial connection, but I guess we could have a use-case where we explicitly accept a 2nd connection.

**Comment by fabioz** [comment]
> I don't know exactly how netcat works but `nc -z` supposedly does not send any data to the port. What I can tell is that `nc` exits immediately after running, so it does not keep the connection open.

It shouldn't leave the connection open, but it does seem like it would do the initial connection to detect it (and then exit right afterward), no? In this case, after anyone does the initial connection, a new client cannot connect anymore as that's not supported in the debugger.

**Comment by fabioz** [comment]
I'm not sure I like this approach... maybe better would be having VSCode itself try a few more times/have a timeout to wait until the connection is up on the server side for this situation?

**Comment by fabioz** [comment]
Great that you found a good alternate solution. Given that, I think it's fine to close this issue.

**Comment by Gaizgrol** [comment]
# TL;DR

If you are using an Alpine based container, run this
```sh
netstat -lnt | grep 5678
```
instead of this
```sh
nc -z localhost 5678
```
for the healthcheck command.

# Detailed answer

I was using `nc -z localhost 5678` as the healthcheck command for the debugger, and I was facing the same problem (debugger couldn't connect after container was healthy)

It turns out that `nc` attempts to establish a connection, so the `--wait-for-client` option discards further attempts and strangely just freezes after that instead of running the code or just exiting the process.

The solution I came up with was using `netstat -lnt | grep 5678` as the healthcheck command, as it just lists the open ports and does not try to establish a connection.
